File: Rectangle.Windows/src/Rectangle.Windows.WinUI/Services/ScreenDetectionService.cs
using System;
using System.Collections.Generic;
using Windows.Win32;
using Windows.Win32.Foundation;
using Windows.Win32.Graphics.Gdi;

namespace Rectangle.Windows.WinUI.Services
{
    /// <summary>
    /// 屏幕检测服务
    /// </summary>
    public class ScreenDetectionService
    {
        private readonly Win32WindowService _win32;

        public ScreenDetectionService(Win32WindowService win32)
        {
            _win32 = win32;
        }

        /// <summary>
        /// 获取所有显示器的工作区域
        /// </summary>
        public List<Core.WindowRect> GetAllWorkAreas()
        {
            var areas = new List<Core.WindowRect>();

            unsafe
            {
                PInvoke.EnumDisplayMonitors(null, null, (hMonitor, hdcMonitor, lprcMonitor, dwData) =>
                {
                    var info = new MONITORINFO { cbSize = (uint)sizeof(MONITORINFO) };
                    if (PInvoke.GetMonitorInfo(hMonitor, &info))
                    {
                        areas.Add(new Core.WindowRect(
                            info.rcWork.left,
                            info.rcWork.top,
                            info.rcWork.right - info.rcWork.left,
                            info.rcWork.bottom - info.rcWork.top));
                    }
                    return true;
                }, 0);
            }

            return areas;
        }

        /// <summary>
        /// 获取鼠标所在的显示器工作区域
        /// </summary>
        public Core.WindowRect GetWorkAreaFromCursor()
        {
            PInvoke.GetCursorPos(out var pt);
            return GetWorkAreaFromPoint(pt.X, pt.Y);
        }

        /// <summary>
        /// 获取指定点所在的显示器工作区域
        /// </summary>
        public Core.WindowRect GetWorkAreaFromPoint(int x, int y)
        {
            unsafe
            {
                var pt = new System.Drawing.Point(x, y);
                var hMonitor = PInvoke.MonitorFromPoint(pt, MONITOR_FROM_FLAGS.MONITOR_DEFAULTTONEAREST);

                var info = new MONITORINFO { cbSize = (uint)sizeof(MONITORINFO) };
                if (PInvoke.GetMonitorInfo(hMonitor, &info))
                {
                    return new Core.WindowRect(
                        info.rcWork.left,
                        info.rcWork.top,
                        info.rcWork.right - info.rcWork.left,
                        info.rcWork.bottom - info.rcWork.top);
                }
            }

            // 默认返回主屏幕
            return GetPrimaryWorkArea();
        }

        /// <summary>
        /// 获取窗口所在的显示器工作区域
        /// </summary>
        public Core.WindowRect GetWorkAreaFromWindow(nint hwnd)
        {
            unsafe
            {
                var hMonitor = PInvoke.MonitorFromWindow(new HWND(hwnd), MONITOR_FROM_FLAGS.MONITOR_DEFAULTTONEAREST);

                var info = new MONITORINFO { cbSize = (uint)sizeof(MONITORINFO) };
                if (PInvoke.GetMonitorInfo(hMonitor, &info))
                {
                    return new Core.WindowRect(
                        info.rcWork.left,
                        info.rcWork.top,
                        info.rcWork.right - info.rcWork.left,
                        info.rcWork.bottom - info.rcWork.top);
                }
            }

            return GetPrimaryWorkArea();
        }

        /// <summary>
        /// 获取主显示器工作区域
        /// </summary>
        public Core.WindowRect GetPrimaryWorkArea()
        {
            unsafe
            {
                var info = new MONITORINFO { cbSize = (uint)sizeof(MONITORINFO) };
                var hMonitor = PInvoke.MonitorFromWindow(new HWND(0), MONITOR_FROM_FLAGS.MONITOR_DEFAULTTOPRIMARY);

                if (PInvoke.GetMonitorInfo(hMonitor, &info))
                {
                    return new Core.WindowRect(
                        info.rcWork.left,
                        info.rcWork.top,
                        info.rcWork.right - info.rcWork.left,
                        info.rcWork.bottom - info.rcWork.top);
                }
            }

            // 如果都失败了，返回一个默认区域
            return new Core.WindowRect(0, 0, 1920, 1080);
        }

        /// <summary>
        /// 获取显示器数量
        /// </summary>
        public int GetDisplayCount()
        {
            int count = 0;
            unsafe
            {
                PInvoke.EnumDisplayMonitors(null, null, (hMonitor, hdcMonitor, lprcMonitor, dwData) =>
                {
                    count++;
                    return true;
                }, 0);
            }
            return count;
        }
    }
}

File: Rectangle.Windows/src/Rectangle.Windows.WinUI.Cpp/Services/ScreenDetectionService.cpp
#include "pch.h"
#include "Services/ScreenDetectionService.h"
#include <Windows.h>
#include <vector>

#pragma comment(lib, "user32.lib")

namespace winrt::Rectangle::Services
{
    ScreenDetectionService::ScreenDetectionService() = default;

    ScreenDetectionService::~ScreenDetectionService() = default;

    std::vector<Core::DisplayInfo> ScreenDetectionService::GetAllDisplays() const
    {
        std::vector<Core::DisplayInfo> displays;

        auto enumCallback = [](HMONITOR hMonitor, HDC hdcMonitor, LPRECT lprcMonitor, LPARAM dwData) -> BOOL
            {
                auto* displays = reinterpret_cast<std::vector<Core::DisplayInfo>*>(dwData);

                MONITORINFO mi = {};
                mi.cbSize = sizeof(MONITORINFO);
                if (GetMonitorInfo(hMonitor, &mi))
                {
                    Core::DisplayInfo info;
                    info.Index = static_cast<int32_t>(displays->size());
                    info.IsPrimary = (mi.dwFlags & MONITORINFOF_PRIMARY) != 0;
                    info.Left = mi.rcMonitor.left;
                    info.Top = mi.rcMonitor.top;
                    info.Right = mi.rcMonitor.right;
                    info.Bottom = mi.rcMonitor.bottom;
                    info.Width = mi.rcMonitor.right - mi.rcMonitor.left;
                    info.Height = mi.rcMonitor.bottom - mi.rcMonitor.top;

                    DISPLAY_DEVICE dd = {};
                    dd.cb = sizeof(DISPLAY_DEVICE);
                    if (EnumDisplayDevices(nullptr, info.Index, &dd, 0))
                    {
                        info.Name = dd.DeviceName;
                    }

                    displays->push_back(info);
                }

                return TRUE;
            };

        EnumDisplayMonitors(nullptr, nullptr, enumCallback, reinterpret_cast<LPARAM>(&displays));

        return displays;
    }

    Core::DisplayInfo ScreenDetectionService::GetPrimaryDisplay() const
    {
        auto displays = GetAllDisplays();
        for (const auto& display : displays)
        {
            if (display.IsPrimary)
            {
                return display;
            }
        }
        return Core::DisplayInfo();
    }

    Core::DisplayInfo ScreenDetectionService::GetDisplayFromPoint(int32_t x, int32_t y) const
    {
        POINT pt = { x, y };
        HMONITOR hMonitor = MonitorFromPoint(pt, MONITOR_DEFAULTTONEAREST);
        if (!hMonitor) return Core::DisplayInfo();

        MONITORINFO mi = {};
        mi.cbSize = sizeof(MONITORINFO);
        if (!GetMonitorInfo(hMonitor, &mi)) return Core::DisplayInfo();

        auto displays = GetAllDisplays();
        for (const auto& display : displays)
        {
            if (display.Left == mi.rcMonitor.left && display.Top == mi.rcMonitor.top)
            {
                return display;
            }
        }

        return Core::DisplayInfo();
    }

    Core::DisplayInfo ScreenDetectionService::GetDisplayFromWindow(int64_t hwnd) const
    {
        HMONITOR hMonitor = MonitorFromWindow(reinterpret_cast<HWND>(hwnd), MONITOR_DEFAULTTONEAREST);
        if (!hMonitor) return Core::DisplayInfo();

        MONITORINFO mi = {};
        mi.cbSize = sizeof(MONITORINFO);
        if (!GetMonitorInfo(hMonitor, &mi)) return Core::DisplayInfo();

        auto displays = GetAllDisplays();
        for (const auto& display : displays)
        {
            if (display.Left == mi.rcMonitor.left && display.Top == mi.rcMonitor.top)
            {
                return display;
            }
        }

        return Core::DisplayInfo();
    }
}
